### PR TITLE
First package version

### PR DESCRIPTION
**Pull request by AntoineAugusti**
https://github.com/opendatateam/udata/pull/2514

image

The plugin uses 2 hooks:
- a hook (`dataset.resource.card.extra-button`) at the resource level on the dataset display page, to add a button to open a modal
- a hook (`base.modals`) at the dataset level, which adds modals to the DOM.

If the dataset has resources with multiple schemas, more than 1 modal will be needed. That's why modals are named based on the resource's ID.

## TODO
- [ ] wait for udata release
- [ ] merge dependency suggestion
- [ ] release 0.1.0 of this package

### CODE REVIEW

File: requirements/install.pip [hunk 1]
```diff
@@ -1 +1 @@
-udata>=2.0.1
+udata==2.1.4.dev19301
```
Comment (abulte) [inline requirements/install.pip:1]:
```suggestion
udata>=2.1.4
```

(tbd after udata's release, just a reminder here)

File: udata_schema_gouvfr/templates/modal.html [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+      <h3>Documentation</h3>
+      <p>
+        Vous pouvez consulter la documentation de ce schéma, découvrir le contexte et les recommandations destinées aux producteurs de données.
+      </p>
+
+      <div>
+        <a
+          href="{{ documentation_urls[resource.id] }}"
+          rel="noopener noreferrer"
```
Comment (abulte) [inline udata_schema_gouvfr/templates/modal.html:20]: 😎

File: udata_schema_gouvfr/templates/modal.html [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+      <h3>Validation</h3>
+      <div class="alert alert-info">
+        <p>
+          La validation d'une ressource par rapport à un schéma consiste à vérifier que la ressource est bien conforme au schéma qu'elle est censée respecter. En cas d'erreurs, un rapport de validation indique les erreurs trouvées : colonnes mal nommées, valeurs non conformes etc.
+        </p>
+      </div>
+
+      <p>
+        data.gouv.fr met à votre disposition l'outil <a href="https://validata.etalab.studio">Validata</a> pour valider vos fichiers par rapport à un schéma et corriger les éventuelles erreurs.
```
Comment (abulte) [inline udata_schema_gouvfr/templates/modal.html]:
```suggestion
        data.gouv.fr met à votre disposition un outil pour valider vos fichiers par rapport à un schéma et corriger les éventuelles erreurs.
```

Le double lien me parait porter à confusion.

File: udata_schema_gouvfr/views.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+from flask import Blueprint
+
+from udata import theme
+from udata.frontend import template_hook
+
+blueprint = Blueprint('schema', __name__, template_folder='templates')
+
+
+def validata_url(resource):
+    base = "https://validata.etalab.studio/table-schema"
```
Comment (abulte) [inline udata_schema_gouvfr/views.py]: Use a setting for this (extensions can declare a `settings.py`)

File: udata_schema_gouvfr/views.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+
+def resource_has_schema(ctx):
+    return ctx.get('resource') and ctx['resource'].schema
+
+
+def dataset_has_schema(ctx):
+    if ctx.get('dataset') is None:
+        return False
+    return any([r.schema is not None for r in ctx['dataset'].resources])
```
Comment (abulte) [inline udata_schema_gouvfr/views.py:30]: 😎

File: udata_schema_gouvfr/views.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+@template_hook('base.modals', when=dataset_has_schema)
+def resource_schema_modal(ctx):
+    dataset = ctx['dataset']
+
+    documentation_urls = {}
+    validation_urls = {}
+    for resource in [r for r in dataset.resources if r.schema]:
+        validation_urls[resource.id] = validata_url(resource)
+        documentation_urls[resource.id] = (
+            f"https://schema.data.gouv.fr/{resource.schema}/latest.html"
```
Comment (abulte) [inline udata_schema_gouvfr/views.py:53]: _maybe_ use a setting for this

**Review by abulte — APPROVED** [review]
Very nice 👌 

Just a general thought: this will perform badly w/ a lot of resources w/ a schema on the same dataset, the DOM will get huge. I don't think this a problem, we just have a to be careful w/ some datasets (LCSQA and DataTourisme). Anyway we need resources pagination on the frontend, but it's out of scope.

**Comment by AntoineAugusti** [comment]
Yeah, you're right to pinpoint the performance issue if the dataset has a large number of resources with a schema. The DOM will be bloated with multiple modals. To solve this, we need to paginate resources or have a modal where we can inject parameters. I'm confident this will not be an issue for now.

I've added a setting for the Validata URL (and a documentation note). I think a setting for the schema.data.gouv.fr URL is superfluous so I chose to let it in the code.

File: requirements/install.pip (hunk 1 above)
Comment (AntoineAugusti) [inline requirements/install.pip:1]:
```suggestion
udata>=2.1.5
```

**Comment by abulte** [comment]
> release 0.1.0 of this package

let's go with 1.0.0 as soon as this PR is tested on demo 💪

**Pull request merged**